src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -829,22 +864,38 @@ function _accumulateInBuffer(
   sampleTime: Milliseconds,
   nextSampleTime: Milliseconds,
   cpuBeforeSample: number | null,
-  cpuAfterSample: number | null
+  cpuAfterSample: number | null,
+  kernelPos?: number
 ) {
   const { rangeStart, xPixelsPerMs, canvasPixelWidth, maxThreadCPUDelta } =
     renderedComponentSettings;
+  const kernelPosNum = kernelPos || 0;
   const sampleCategoryStartTime = (prevSampleTime + sampleTime) / 2;
   const sampleCategoryEndTime = (sampleTime + nextSampleTime) / 2;
   let sampleCategoryStartPixel =
-    (sampleCategoryStartTime - rangeStart) * xPixelsPerMs;
+    (sampleCategoryStartTime - rangeStart) * xPixelsPerMs - kernelPosNum;
   let sampleCategoryEndPixel =
-    (sampleCategoryEndTime - rangeStart) * xPixelsPerMs;
+    (sampleCategoryEndTime - rangeStart) * xPixelsPerMs - kernelPosNum;
   sampleCategoryStartPixel = Math.max(0, sampleCategoryStartPixel);
   sampleCategoryEndPixel = Math.min(
     canvasPixelWidth - 1,
     sampleCategoryEndPixel
   );
-  const samplePixel = (sampleTime - rangeStart) * xPixelsPerMs;
+  let samplePixel = (sampleTime - rangeStart) * xPixelsPerMs - kernelPosNum;
+  if (kernelPos) {
+    // Clamp the pixels to smoothing kernel in case there is a kernel position.
+    sampleCategoryStartPixel = clamp(
+      sampleCategoryStartPixel,
+      0,
+      SMOOTHING_KERNEL.length - 1
+    );
+    sampleCategoryEndPixel = clamp(
+      sampleCategoryEndPixel,
+      0,
+      SMOOTHING_KERNEL.length - 1
+    );
+    samplePixel = clamp(samplePixel, 0, SMOOTHING_KERNEL.length - 1);
+  }
```

Comment:
Actually, activity graph drawing will not be slowed down because these clampings only run if `kernelPos` is not undefined (which is undefined for drawing path, it's only non-undefined for hit testing case). 
We were also already doing these clampings for hit testing path. We only had to add one more clamp call. It was only for start and end before. Now it's for start, sample time and end.  But let me try changing canvasPixelWidth with percentageBuffer.length if that makes things simpler.

Reply:
Oh, true, so the only extra cost on the drawing path would be to check if kernelPos is truthy. Ok, so then performance is probably not the right argument for me to make.
However, I do think my suggested modification would simplify the code.